### PR TITLE
Avoid creating a task to do DNS resolution if there is no throttle

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

Avoid creating a task to do dns resolution if there is no throttle

When doing some debugging, I noticed that a large amount of tasks created on a Home Assistant instance were `TCPConnector._resolve_host` for cases where the host was an ip address or already in the cache.  On this instance, it reduced the number of `_resolve_host` tasks by 99%. Previously I observed ~500 per minute.

## Are there changes in behavior for the user?

There is a small performance improvement on cache hit and when the host is an ip address.  Since requests no longer have to suspend on cache hit this can reduce latency of requests.  Scheduling and awaiting a task on the event loop can take 2 order of magnitude more wall clock time vs running a simple function.

## Is it a substantial burden for the maintainers to support this?

no

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES/` folder
  * name it `<issue_or_pr_num>.<type>.rst` (e.g. `588.bugfix.rst`)
  * if you don't have an issue number, change it to the pull request
    number after creating the PR
    * `.bugfix`: A bug fix for something the maintainers deemed an
      improper undesired behavior that got corrected to match
      pre-agreed expectations.
    * `.feature`: A new behavior, public APIs. That sort of stuff.
    * `.deprecation`: A declaration of future API removals and breaking
      changes in behavior.
    * `.breaking`: When something public is removed in a breaking way.
      Could be deprecated in an earlier release.
    * `.doc`: Notable updates to the documentation structure or build
      process.
    * `.packaging`: Notes for downstreams about unobvious side effects
      and tooling. Changes in the test invocation considerations and
      runtime assumptions.
    * `.contrib`: Stuff that affects the contributor experience. e.g.
      Running tests, building the docs, setting up the development
      environment.
    * `.misc`: Changes that are hard to assign to any of the above
      categories.
  * Make sure to use full sentences with correct case and punctuation,
    for example:
    ```rst
    Fixed issue with non-ascii contents in doctest text files
    -- by :user:`contributor-gh-handle`.
    ```

    Use the past tense or the present tense a non-imperative mood,
    referring to what's changed compared to the last released version
    of this project.

### CODE REVIEW

File: tests/test_connector.py
Comment (bdraco) [inline tests/test_connector.py:1036]: These tests all rely on the event loop overhead introduced by the task. Without the task we need to let it run a few more iterations because the test gets to run sooner now

Comment (Dreamsorcerer) [reply tests/test_connector.py:1036]: I'd probably just put a delay in (0.01 or something) if it's not guaranteed to be done in the next iteration.

**Review by Dreamsorcerer — APPROVED** [review]

File: aiohttp/connector.py
Comment (webknjaz) [inline aiohttp/connector.py]: PEP 257 requires a sentence in imperative mood, in docstring title lines: https://peps.python.org/pep-0257/#one-line-docstrings

File: aiohttp/connector.py
Comment (webknjaz) [inline aiohttp/connector.py]: Maybe, call it a `resolved_host_task`?

File: aiohttp/connector.py
Comment (webknjaz) [inline aiohttp/connector.py:871]: No need to quote the annotation since there's no py310-only syntax here:
```suggestion
            def drop_exception(fut: asyncio.Future[List[Dict[str, Any]]]) -> None:
```

Comment (bdraco) [reply aiohttp/connector.py:871]: I'm not sure asyncio.Future is sub-scriptable on older python.  Will drop it and see if the CI complains

File: aiohttp/connector.py
Comment (bdraco) [reply aiohttp/connector.py]: I moved the code from below and didn't think about renaming it.  Sounds good to me.

File: aiohttp/connector.py
Comment (bdraco) [reply aiohttp/connector.py]: 2367798d26f4d8928e618976f8d231e2160ce16b

**Comment by bdraco** [comment]
How do we feel about backporting this to 3.9.  HA is having a lot of stability problems with too many tasks, and this would help a lot in reducing them.

File: tests/test_connector.py
Comment (bdraco) [reply tests/test_connector.py:1036]: 8609c35605022650ca07d04fdedd76f708b7b372

I had to leave the two in test_tcp_connector_dns_throttle_requests_cancelled_when_close since it failed otherwise

File: aiohttp/connector.py
Comment (Dreamsorcerer) [reply aiohttp/connector.py:871]: Just tried, it's not.

Comment (bdraco) [reply aiohttp/connector.py:871]: restored quoting 38e0649023c32bcb2c986a0aa744f5ad17799c32

File: tests/test_connector.py
Comment (bdraco) [reply tests/test_connector.py:1036]: I ended up reverting them since the test was only reliable as when it waited X iterations of the event loop instead of time

File: aiohttp/connector.py
Comment (webknjaz) [reply aiohttp/connector.py:871]: Ah, I didn't realize that.

File: aiohttp/connector.py
Comment (webknjaz) [inline aiohttp/connector.py]:
```suggestion
        """Resolve host and return list of addresses."""
```

**Comment by webknjaz** [comment]
> How do we feel about backporting this to 3.9. HA is having some stability problems with too many tasks, and this would help a lot in reducing them.

I think this can easily be declared a bugfix since (1) it's not a new feature (meaning no grounds for requiring a minor version bump) and (2) it fixes an undesired behavior.

File: aiohttp/connector.py
Comment (bdraco) [reply aiohttp/connector.py]: thanks

File: CHANGES/8163.bugfix.rst
Comment (webknjaz) [inline CHANGES/8163.bugfix.rst]: Check out the recently improved guidelines: https://docs.aiohttp.org/en/latest/contributing.html#changelog-update / https://github.com/aio-libs/aiohttp/tree/master/CHANGES#readme.

Perhaps, something like this would be easier comprehend for the end-user audience:
```suggestion
Improved the not-throttled DNS resolution performance
-- by :user:`bdraco`.

This is achieverd by avoiding an :mod:`asyncio` task creation
in this case.
```

In case, that's hard to follow, here's somebody else's attempt to encode good release notes writing style: https://redhat-documentation.github.io/supplementary-style-guide/#release-notes (it doesn't fully correspond to our needs, though).

File: CHANGES/8163.bugfix.rst
Comment (bdraco) [inline CHANGES/8163.bugfix.rst]:
```suggestion
This is achieved by avoiding an :mod:`asyncio` task creation
```

File: CHANGES/8163.bugfix.rst
Comment (bdraco) [reply CHANGES/8163.bugfix.rst]: Works for me. Thanks. It really helps to get another opinion on how to describe the change since its hard to guess how someone who didn't write the change will want to consume it.... I am not the user problem.

Comment (webknjaz) [reply CHANGES/8163.bugfix.rst]: Yep, it's hard for everyone which is why I'm failing to come up with better guidelines :)

I just attempted rephrasing what you posted in the PR, that's all my context now.

Comment (Dreamsorcerer) [reply CHANGES/8163.bugfix.rst]: "Performance: Avoided creating tasks unnecessarily during DNS resolution."

would cover it for me. I don't even know what throttled means in this context.

File: CHANGES/8163.bugfix.rst
Comment (bdraco) [inline CHANGES/8163.bugfix.rst]:
```suggestion
Improved the DNS resolution performance on cache hit
```

File: CHANGES/8163.bugfix.rst
Comment (bdraco) [reply CHANGES/8163.bugfix.rst]: I adjusted it again to remove the word `throttled`

Comment (webknjaz) [reply CHANGES/8163.bugfix.rst]: That's too low-level. Throttling refers to limiting the number of simultaneous connections to resolve the same hostname, it seems.

Comment (Dreamsorcerer) [reply CHANGES/8163.bugfix.rst]: > Throttling refers to limiting the number of simultaneous connections to resolve the same hostname, it seems.

Right, to me that's just sensible caching (e.g. If you just added `@alru_cache` to the function it would also do that).

Comment (webknjaz) [reply CHANGES/8163.bugfix.rst]: Brilliant! That's much better!

**Review by webknjaz — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks for the reviews.  I'll merge it tomorrow or later tonight after some sleep + recheck.

**Comment by bdraco** [comment]
Still looks good, functional retest is good.

**Pull request merged**
